cirq-core/cirq/value/condition.py

```diff
@@ -0,0 +1,168 @@
+    """
+
+    key: 'cirq.MeasurementKey'
+
+    @property
+    def keys(self):
+        return (self.key,)
+
+    def with_keys(self, *keys: 'cirq.MeasurementKey'):
+        assert len(keys) == 1
+        return KeyCondition(keys[0])
+
+    def __str__(self):
+        return str(self.key)
+
+    def resolve(self, measurements: Mapping[str, Sequence[int]]) -> bool:
+        key = str(self.key)
+        if key not in measurements:
+            raise ValueError(f'Measurement key {key} missing when testing classical control')
+        return any(measurements[key])
+
+    def _json_dict_(self):
+        return json_serialization.dataclass_json_dict(self)
+
+    @property
+    def qasm(self):
+        return f'm_{self.key}!=0'
+
+
+@dataclasses.dataclass(frozen=True)
+class SympyCondition(Condition):
+    """A classical control condition based on a sympy expression.
+
+    This condition resolves to True iff the sympy expression resolves to a
+    Truthy value when the measurement keys are substituted in as the free
+    variables.
+
+    To account for the fact that measurement key strings can contain characters
+    not allowed in sympy variables, we use x0..xN for the free variables and
+    substitute the measurement keys in by order. For instance a condition with
+    `expr='x0 > x1', keys=['0:A', '0:B']` would resolve to True iff key `0:A`
+    is greater than key `0:B`.
+
+    The `cirq.parse_sympy_condition` function automates setting this up
+    correctly. To create the above expression, one would call
+    `cirq.parse_sympy_condition('{0:A} > {0:B}')`.
+    """
+
+    expr: sympy.Expr
+    control_keys: Tuple['cirq.MeasurementKey', ...]
+
+    @property
+    def keys(self):
+        return self.control_keys
+
+    def with_keys(self, *keys: 'cirq.MeasurementKey'):
+        assert len(keys) == len(self.control_keys)
+        return dataclasses.replace(self, control_keys=keys)
+
+    def __str__(self):
+        replacements = {f'x{i}': f'{{{str(key)}}}' for i, key in enumerate(self.control_keys)}
+
+        class CustomCodePrinter(sympy.printing.StrPrinter):
+            def _print_Symbol(self, expr):
+                return replacements[expr.name]
+
+        return CustomCodePrinter().doprint(self.expr)
+
+    def resolve(self, measurements: Mapping[str, Sequence[int]]) -> bool:
+        missing = [str(k) for k in self.keys if str(k) not in measurements]
+        if missing:
+            raise ValueError(f'Measurement keys {missing} missing when testing classical control')
+
+        def value(k):
+            return digits.big_endian_bits_to_int(measurements[str(k)])
+
+        replacements = {f'x{i}': value(k) for i, k in enumerate(self.keys)}
+        return bool(self.expr.subs(replacements))
+
+    def _json_dict_(self):
+        return json_serialization.dataclass_json_dict(self)
+
+    @property
+    def qasm(self):
+        raise NotImplementedError()
+
+
+def parse_sympy_condition(s: str) -> 'cirq.SympyCondition':
```

Comment:
[Optional] Part of me wants this to be a static method like `SympyCondition.from_expr()`, though of course that doesn't work nicely for `parse_condition` below...what do you think?

Reply:
Not opposed, but it strictly goes against https://google.github.io/styleguide/pyguide.html#s2.17-function-and-method-decorators. Since it sounds like you're on the fence, I think best to stick with the style guide. (Though granted, we've broken the rule all over the place already, so).